### ISSUE TITLE
gRPC does not reconnect after broker restart

### ISSUE THREAD

**Issue by jwulf** [creditsenseau/zeebe-client-node-js#66]
To emulate this, start a  broker, then a worker.

Kill the broker, then restart it.

Previously connected workers do not reconnect. With a long-polling worker with `loglevel` set to `DEBUG` you can see the failure retrying continuously.

According to these, the underlying gRPC library should reconnect automagically:

- https://www.oipapio.com/question-4759444
- https://stackoverflow.com/questions/42256810/how-can-i-change-grpcs-reconnection-behaviour-in-the-node-js-implementation
- https://github.com/grpc/grpc/issues/11411#issuecomment-308899967
- https://github.com/grpc/grpc/issues/11411#issuecomment-306333733
- https://grpc.github.io/grpc/cpp/md_doc_connectivity-semantics-and-api.html

Maybe something else needs to be configured in the optional third parameter to Client in GRPCClient?
https://grpc.github.io/grpc/core/group__grpc__arg__keys.html#gad7d9d143858d8f5e138cf704b0082973

**Comment by jwulf** [comment]
Moar:

* https://stackoverflow.com/questions/43489277/grpc-looses-connection-upon-server-restart
* https://github.com/grpc/grpc/issues/9767
https://github.com/ncteisen/proposal/blob/4d61d04e3db298b1c2503e1ab0245f16e1b86967/A6.md#transparent-retries
* https://godoc.org/google.golang.org/grpc

**Comment by jwulf** [comment]
Instrumenting the worker like this (should go in GRPCClient, because it is shared by workers):

```
	private handleGrpcError = (err: any) => {
		const connectivityState = [
			'IDLE',
			'CONNECTING',
			'READY',
			'TRANSIENT_FAILURE',
			'SHUTDOWN',
		]
		const gRPC = this.gRPCClient.client
		const state = gRPC.getChannel().getConnectivityState(false)

		const deadline = new Date().setSeconds(new Date().getSeconds() + 30)
		gRPC.getChannel().watchConnectivityState(state, deadline, error => {
			this.gRPCRetryCount++
			if (error) this.logger.error({ error })
			const newState = gRPC.getChannel().getConnectivityState(false)
			this.logger.log(
				`GRPC Channel State: ${connectivityState[newState]}`
			)
			this.logger.info(`gRPC Retry count: ${this.gRPCRetryCount}`)
		})
		if (!this.errored) {
			this.logger.error(`GRPC Channel State: ${connectivityState[state]}`)

			if (this.onConnectionErrorHandler) {
				this.onConnectionErrorHandler(err)
				this.errored = true
			} else {
				this.logger.error(`GRPC ERROR: ${err.message}`)
				this.errored = true
			}
		}
		if (this.longPoll) {
			this.logger.debug(`gRPC Error ${chalk.red(err.details)}`)
			this.logger.debug('Calling longPollLoop()')
			this.longPollLoop()
		}
	}
```

**Comment by jwulf** [comment]
OK, I got this working, with logic inside the ZBWorker. Now I have to move that logic down into the GRPCClient, so it is handled per gRPC channel, rather than per worker.

**Comment by jwulf** [comment]
Closed by #68
